### ISSUE TITLE
Node left into draining state in multi node set up

### ISSUE THREAD

**Issue by gkokolatos** [citusdata/pg_auto_failover#704]
Hi team,

we are noticing a behaviour which is consistently reproducible since the introduction of multi node set up. The checks were performed against postgres versions 11 and 12. We are wondering if this is the expected behaviour.

Assume a three node setup with data nodes named node[1-3], and one monitor node.
Assume that node1 is primary and nodes node2 and node3 are secondary. The system is healthy.
Bring the primary down and wait until a new primary is elected, e.g. node2. Bring down that new primary node2.
Wait until the monitor has registered the state and reacted. Now node3 should be in wait_primary state.
Bring back up node2. Now node2 is stuck into draining state. What is interesting is that node2's db is down
and the system does not seem to react to it. It stays permanently in that state.

Let us show the states of the cluster (pg_autoctl show state against the monitor node) after stopping node1.

```
  Name |  Node |      Host:Port |       LSN |   Connection |       Current State |      Assigned State
-------+-------+----------------+-----------+--------------+---------------------+--------------------
node_1 |     1 | 10.0.2.15:5001 | 0/5000108 | read-write ! |             primary |             demoted
node_2 |     2 | 10.0.2.15:5002 | 0/5000410 |   read-write |             primary |             primary
node_3 |     3 | 10.0.2.15:5003 | 0/5000410 |    read-only |           secondary |           secondary
```

This is all good and expected.

Let us show the states of the cluster after stopping node2

```
  Name |  Node |      Host:Port |       LSN |   Connection |       Current State |      Assigned State
-------+-------+----------------+-----------+--------------+---------------------+--------------------
node_1 |     1 | 10.0.2.15:5001 | 0/5000108 | read-write ! |             primary |             demoted
node_2 |     2 | 10.0.2.15:5002 | 0/5000410 | read-write ! |             primary |            draining
node_3 |     3 | 10.0.2.15:5003 | 0/5000638 |   read-write |        wait_primary |        wait_primary
```

This state does not change and node_2 remains in draining Assigned state.
The registered events at this point (pg_autoctl show events against the monitor node) show

```
                    Event Time |   Node |       Current State |      Assigned State | Comment
-------------------------------+--------+---------------------+---------------------+-----------
 2021-05-07 14:20:20.097981+00 |    0/3 |           secondary | report_lsn | Setting goal state of node 3 "node_3" (10.0.2.15:5003) to report_lsn to find the failover candidate
 2021-05-07 14:20:26.482868+00 |    0/3 |          report_lsn | report_lsn | New state is reported by node 3 "node_3" (10.0.2.15:5003) with LSN 0/5000480: report_lsn
 2021-05-07 14:20:26.482868+00 |    0/3 |          report_lsn | prepare_promotion | Setting goal state of node 3 "node_3" (10.0.2.15:5003) to prepare_promotion after node 1 "node_1" (10.0.2.15:5001) became unhealthy and 1 nodes reported their LSN position.
 2021-05-07 14:20:27.556603+00 |    0/3 |   prepare_promotion | prepare_promotion | New state is reported by node 3 "node_3" (10.0.2.15:5003): "prepare_promotion"
 2021-05-07 14:20:27.556603+00 |    0/3 |   prepare_promotion | stop_replication | Setting goal state of node 1 "node_1" (10.0.2.15:5001) to demote_timeout and node 3 "node_3" (10.0.2.15:5003) to stop_replication after node 3 "node_3" (10.0.2.15:5003) converged to prepare_promotion.
 2021-05-07 14:20:27.556603+00 |    0/1 |             primary | demote_timeout | Setting goal state of node 1 "node_1" (10.0.2.15:5001) to demote_timeout and node 3 "node_3" (10.0.2.15:5003) to stop_replication after node 3 "node_3" (10.0.2.15:5003) converged to prepare_promotion.
 2021-05-07 14:20:28.826102+00 |    0/3 |    stop_replication | stop_replication | New state is reported by node 3 "node_3" (10.0.2.15:5003): "stop_replication"
 2021-05-07 14:20:57.977099+00 |    0/3 |    stop_replication | wait_primary | Setting goal state of node 3 "node_3" (10.0.2.15:5003) to wait_primary and node 1 "node_1" (10.0.2.15:5001) to demoted after the demote
timeout expired.
 2021-05-07 14:20:57.977099+00 |    0/1 |             primary | demoted | Setting goal state of node 3 "node_3" (10.0.2.15:5003) to wait_primary and node 1 "node_1" (10.0.2.15:5001) to demoted after the demote timeout
expired.
  2021-05-07 14:20:58.02127+00 |    0/3 |        wait_primary | wait_primary | New state is reported by node 3 "node_3" (10.0.2.15:5003): "wait_primary"
```

What strikes a bit as curious is that no events for node 0/2 are shown.

Now bring node_2 back up again.
The cluster stays for ever in the following state:
``` 
  Name |  Node |      Host:Port |       LSN |   Connection |       Current State |      Assigned State
-------+-------+----------------+-----------+--------------+---------------------+--------------------
node_1 |     1 | 10.0.2.15:5001 | 0/5000108 | read-write ! |             primary |             demoted
node_2 |     2 | 10.0.2.15:5002 | 0/5000410 |       none ! |            draining |            draining
node_3 |     3 | 10.0.2.15:5003 | 0/5000638 |   read-write |        wait_primary |        wait_primary
```

The registered events are:
```
                    Event Time |   Node |       Current State |      Assigned State | Comment
-------------------------------+--------+---------------------+---------------------+-----------
 2021-05-07 14:20:26.482868+00 |    0/3 |          report_lsn | report_lsn | New state is reported by node 3 "node_3" (10.0.2.15:5003) with LSN 0/5000480: report_lsn
 2021-05-07 14:20:26.482868+00 |    0/3 |          report_lsn | prepare_promotion | Setting goal state of node 3 "node_3" (10.0.2.15:5003) to prepare_promotion after node 1 "node_1" (10.0.2.15:5001) became unhealthy and 1 nodes reported their LSN position.
 2021-05-07 14:20:27.556603+00 |    0/3 |   prepare_promotion | prepare_promotion | New state is reported by node 3 "node_3" (10.0.2.15:5003): "prepare_promotion"
 2021-05-07 14:20:27.556603+00 |    0/3 |   prepare_promotion | stop_replication | Setting goal state of node 1 "node_1" (10.0.2.15:5001) to demote_timeout and node 3 "node_3" (10.0.2.15:5003) to stop_replication after node 3 "node_3" (10.0.2.15:5003) converged to prepare_promotion.
 2021-05-07 14:20:27.556603+00 |    0/1 |             primary | demote_timeout | Setting goal state of node 1 "node_1" (10.0.2.15:5001) to demote_timeout and node 3 "node_3" (10.0.2.15:5003) to stop_replication after node 3 "node_3" (10.0.2.15:5003) converged to prepare_promotion.
 2021-05-07 14:20:28.826102+00 |    0/3 |    stop_replication | stop_replication | New state is reported by node 3 "node_3" (10.0.2.15:5003): "stop_replication"
 2021-05-07 14:20:57.977099+00 |    0/3 |    stop_replication | wait_primary | Setting goal state of node 3 "node_3" (10.0.2.15:5003) to wait_primary and node 1 "node_1" (10.0.2.15:5001) to demoted after the demote timeout expired.
 2021-05-07 14:20:57.977099+00 |    0/1 |             primary | demoted | Setting goal state of node 3 "node_3" (10.0.2.15:5003) to wait_primary and node 1 "node_1" (10.0.2.15:5001) to demoted after the demote timeout expired.
  2021-05-07 14:20:58.02127+00 |    0/3 |        wait_primary | wait_primary | New state is reported by node 3 "node_3" (10.0.2.15:5003): "wait_primary"
 2021-05-07 14:24:38.105043+00 |    0/2 |            draining | draining | New state is reported by node 2 "node_2" (10.0.2.15:5002): "draining"
```

The logs show:

```
Started pg_autoctl postgres service with pid 364603
Started pg_autoctl node-active service with pid 364604
 ./bin/pg_autoctl do service postgres --pgdata /tmp/node2 -v
keeper has been successfully initialized.
 ./bin/pg_autoctl do service node-active --pgdata /tmp/node2 -v
Reloaded the new configuration from ".config/pg_autoctl/tmp/node2/pg_autoctl.cfg"
pg_autoctl service is running, current state is "primary"
Failed to fetch current replication properties from standby node: no standby connected in pg_stat_replication.
HINT: check pg_autoctl and Postgres logs on standby nodes
Failed to update the keeper's state from the local PostgreSQL instance.
Fetched current list of 2 other nodes from the monitor to update HBA rules, including 2 changes.
Ensuring HBA rules for node 1 "node_1" (10.0.2.15:5001)
Ensuring HBA rules for node 3 "node_3" (10.0.2.15:5003)
Monitor assigned new state "draining"
FSM transition from "primary" to "draining": A failover occurred, stopping writes
Transition complete: current state is now "draining"
Updated the keeper's state from the local PostgreSQL instance, which is not running
pg_autoctl managed to ensure current state "draining": PostgreSQL is not running
```

Is the above the expected behaviour?

Thank you in advance for your time.

P.D. Bringing node1 up, will not change the state of node2.

**Comment by DimCitus** [comment]
Hi @gkokolatos ; thanks for reporting your findings in a detailed way, so that we can reproduce the issue. I just followed the steps outlined here and could reproduce the problem. This is not expected behaviour. My first analysis is that when a node has been assigned the DRAINING state (at failover trigger time), there is only one way in the monitor Group FSM to then move forward:

``` c
	/*
	 * when drain time expires or primary reports it's drained:
	 *  draining -> demoted
	 */
	if (IsCurrentState(activeNode, REPLICATION_STATE_STOP_REPLICATION) &&
		(IsCurrentState(primaryNode, REPLICATION_STATE_DEMOTE_TIMEOUT) ||
		 IsDrainTimeExpired(primaryNode)))
	{
		char message[BUFSIZE];

		LogAndNotifyMessage(
			message, BUFSIZE,
			"Setting goal state of " NODE_FORMAT
			" to wait_primary and " NODE_FORMAT
			" to demoted after the demote timeout expired.",
			NODE_FORMAT_ARGS(activeNode),
			NODE_FORMAT_ARGS(primaryNode));

		/* node is now taking writes */
		AssignGoalState(activeNode, REPLICATION_STATE_WAIT_PRIMARY, message);

		/* done draining, node is presumed dead */
		AssignGoalState(primaryNode, REPLICATION_STATE_DEMOTED, message);

		return true;
	}
```

I think we should add to that list of transitions a transition from DRAINING to DEMOTE_TIMEOUT when `IsDrainTimeExpired(primaryNode)`, without other conditions on other nodes. That should be done later in the Group FSM code, first matching situations that are in the middle of a failover, where we then can have one less round-trip to make progress forward.

Do you want to have a try at it, including adding a new test case that reproduces the list of steps you outlined in the issue?

**Comment by gkokolatos** [comment]
> Do you want to have a try at it, including adding a new test case that reproduces the list of steps you outlined in the issue?

Thanks you for your prompt reply and thoughts @DimCitus. I would be glad to try out your suggestion and submit a PR in the coming days.

**Comment by DimCitus** [comment]
After more reading of the code, I think the problem is a little different. What happens is that we have now two nodes in states that are covered by `GetPrimaryOrDemotedNodeInGroup` (one DEMOTED, one DRAINING), and the code is prepared to only deal with a single of those at any time. My current best guess is then that we should have a special case for nodes in DRAINING/DEMOTED/DEMOTE_TIMEOUT state in particular when they're not running and when their assigned state and reported state are different.

``` c
	/*
	 * We separate out the FSM for the primary server, because that one needs
	 * to loop over every other node to take decisions. That induces some
	 * complexity that is best managed in a specialized function.
	 */
	if (IsInPrimaryState(activeNode))
	{
		return ProceedGroupStateForPrimaryNode(activeNode);
	}

	AutoFailoverNode *primaryNode =
		GetPrimaryOrDemotedNodeInGroup(formationId, groupId);
```

We might want to revisit the whole `GetPrimaryOrDemotedNodeInGroup` idea, or just prepend a special case to deal with one-or-more nodes being assigned one of the draining/demoted/demote_timeout states.

**Comment by gkokolatos** [comment]
@DimCitus As expected you were spot on with your deduction of `GetPrimaryOrDemotedNodeInGroup` as the culprit.

Find a PR to address the issue.

**Comment by DimCitus** [comment]
Fixed in #706
